src/Analyzers/CSharp/Analyzers/UseSystemThreadingLock/CSharpUseSystemThreadingLockDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,268 @@
+                } field)
+            {
+                continue;
+            }
+
+            var fieldSyntaxTree = fieldSyntaxReference.SyntaxTree;
+            if (fieldSyntaxTree != currentSyntaxTree)
+            {
+                currentSyntaxTree = fieldSyntaxTree;
+
+                currentOption = context.GetCSharpAnalyzerOptions(currentSyntaxTree).PreferSystemThreadingLock;
+
+                // Ignore this field if it is is in a file that should be skipped.
+                if (!currentOption.Value || ShouldSkipAnalysis(currentSyntaxTree, context.Options, context.Compilation.Options, currentOption.Notification, cancellationToken))
+                    continue;
+            }
+
+            if (fieldSyntaxReference.GetSyntax(cancellationToken) is not VariableDeclaratorSyntax variableDeclarator)
+                continue;
+
+            // For simplicity, only offer this for fields with a single declarator.
+            if (variableDeclarator.Parent is not VariableDeclarationSyntax { Parent: FieldDeclarationSyntax fieldDeclaration, Variables.Count: 1 })
+                return;
+
+            // Looks like something that could be converted to a System.Threading.Lock if we see that this field is used
+            // in a compatible fashion.
+            fieldsArray.Add((field, fieldDeclaration, currentOption!));
+        }
+
+        if (fieldsArray.Count == 0)
+            return;
+
+        // The set of fields we think could be converted to `System.Threading.Lock` from `object`.
+        //
+        // 'wasLocked' tracks whether or not we saw this field used in a `lock (obj)` statement.  If not, we do not want
+        // to convert this as the user wasn't using this as a lock.  Note: we can consider expanding the set of patterns
+        // we detect (like Monitor.Enter + Monitor.Exit) if we think it's worthwhile.
+        //
+        // Note: both this dictionary is written to concurrently in the analysis pass below.  This is safe as we never
+        // are actually mutating the dictionary itself (we're not adding/removing items).  We're just getting a
+        // reference to its tuple value and overwriting a bool in place within that tuple with the new value.  And we
+        // only move hte value in one direction.  For example 'canUse' only moved from 'true' to 'false' and 'wasLocked'
+        // only moves the value from 'false' to 'true'.
+        var potentialLockFields = new SegmentedDictionary<IFieldSymbol, (FieldDeclarationSyntax fieldDeclaration, CodeStyleOption2<bool> option, bool canUse, bool wasLocked)>();
+
+        foreach (var (field, fieldDeclaration, option) in fieldsArray)
+            potentialLockFields[field] = (fieldDeclaration, option, canUse: true, wasLocked: false);
+
+        // Register a syntax node callback to ensure the field's initializer is either missing, or is only instantiating
+        // the field with a new object.
+        context.RegisterSyntaxNodeAction(context =>
```

Comment:
💡 Can use an operation action for OperationKind.FieldInitializer? That would allow lookup in `potentialLockFields` instead of looping through `fieldsArray`

Reply:
nice.  i didn't know about FieldInitializer.  that mkes it much cleaner.